libp2p/protocols/secure/noise.nim

```diff
@@ -0,0 +1,575 @@
+    
+  var
+    libp2pProof = initProtoBuffer()
+    handshakeRes: HandshakeResult
+
+  libp2pProof.write(initProtoField(1, p.localPublicKey))
+  libp2pProof.write(initProtoField(2, signedPayload))
+  # data field also there but not used!
+  libp2pProof.finish()
+
+  if initiator:
+    handshakeRes = await handshakeXXOutbound(p, conn, libp2pProof)
+  else:
+    handshakeRes = await handshakeXXInbound(p, conn, libp2pProof)
+
+  var
+    remoteProof = initProtoBuffer(handshakeRes.remoteP2psecret)
+    remotePubKey: PublicKey
+    remoteSig: Signature
+  if remoteProof.getValue(1, remotePubKey) <= 0:
+    raise newException(NoiseHandshakeError, "Failed to deserialize remote public key.")
+  if remoteProof.getValue(2, remoteSig) <= 0:
+    raise newException(NoiseHandshakeError, "Failed to deserialize remote public key.")
+
+  let verifyPayload = PayloadString.getBytes & handshakeRes.rs.getBytes
+  if not remoteSig.verify(verifyPayload, remotePubKey):
+    raise newException(NoiseHandshakeError, "Noise handshake signature verify failed.")
+  else:
+    trace "Remote signature verified"
+ 
+  if initiator and not isNil(conn.peerInfo):
+    let pid = PeerID.init(remotePubKey)
+    if not conn.peerInfo.peerId.validate():
+      raise newException(NoiseHandshakeError, "Failed to validate peerId.")
+    if pid != conn.peerInfo.peerId:
+      raise newException(NoiseHandshakeError, "Noise handshake, peer infos don't match! " & $pid & " != " & $conn.peerInfo.peerId)
+
+  var secure = new NoiseConnection
+  secure.stream = conn
+  secure.closeEvent = newAsyncEvent()
+  secure.peerInfo = PeerInfo.init(remotePubKey)
+  if initiator:
+    secure.readCs = handshakeRes.cs2
+    secure.writeCs = handshakeRes.cs1
+  else:
+    secure.readCs = handshakeRes.cs1
+    secure.writeCs = handshakeRes.cs2
+
+  trace "Noise handshake completed!"
+
+  return secure
+ 
+method init*(p: Noise) {.gcsafe.} =
+  trace "Noise init called"
+ 
+  proc handle(conn: Connection, proto: string) {.async, gcsafe.} =
+    trace "handling connection", proto
+    try:
+      let
+        sconn = await p.handshake(conn, false)
+      asyncCheck sconn.startLifetime()
+      trace "connection secured"
+    except CatchableError as ex:
+      if not conn.closed():
+        warn "securing connection failed", msg = ex.msg
+        await conn.close()
+
+  p.codec = NoiseCodec
+  p.handler = handle
+  
+method secure*(p: Noise, conn: Connection, outgoing: bool): Future[Connection] {.async, gcsafe.} =
+  try:
+    let
+      sconn = await p.handshake(conn, outgoing)
+      securedStream = await sconn.startLifetime()
+    securedStream.peerInfo = sconn.peerInfo
+    return securedStream
+  except CatchableError as ex:
+     warn "securing connection failed", msg = ex.msg
+     if not conn.closed():
+       await conn.close()
+  
+proc newNoise*(privateKey: PrivateKey; commonPrologue: seq[byte] = @[]): Noise =
+  new result
```

Comment:
`Noise(a: b, c:d)` is a safer style that carries fewer risks of using a partially initialized object - also, we often create a `func init(T: type Noise, params): T = T(...)` to initialize things genertically - this is as close to a constructor as you'll get.

Reply:
Adopted this style from `secio.nim` as well.

Reply:
@dryajov I guess I could change Secio as well but that should go into another branch probably no?

Reply:
Yes, there is a general move towards  the `MyType(a: b, c:d)` style and we should refactor everywhere, but we haven't got there yet.

Reply:
Will do that after merge cos secio needs it too0